text/0009-multilink_command_handling.md

```diff
@@ -0,0 +1,36 @@
+  * Start date: 2018-05-21
+  * Contributors: Achermann Florian, <acfloria@ethz.ch>
+  * Related issues: 
+  
+# Summary
+
+The goal of this document is to present a design on how to handle commands which are sent over different links with significantly different latencies from the ground control station to a vehicle.
+  
+# Motivation
+
+When transmitting commands over multiple links with significantly different latencies to a single vehicle the order in which the commands are sent might not be equal to the order in which they are received. This can lead to an undesired response of the vehicle especially when using commands such as MAV_CMD_DO_CHANGE_ALTITUDE, MAV_CMD_DO_SET_MODE, or MAV_CMD_DO_REPOSITION.
+
+# Detailed Design
+
+A new type of command is created: COMMAND_STAMPED. This command is basically a COMMAND_LONG with additional metadata:
+  * (uint64) vehicle_time: The vehicle time since boot in ms
+  * (uint64) utc_time: The seconds elapsed since Jan 01 1970 (UTC)
+
+The vehicle_time field is filled in with the value of the timestamp field of the last received vehicle time on the groundstation (SYSTEM_TIME or HIGH_LATENCY2). This indicates on which vehicle status the command is sent. On the groundstation the last vehicle is reset to 0 after some time not receiving an update to avoid that any sent command is rejected.
+
+The utc_time provides a reference between different flights. Which is necessary as for example Satellite Communication messages can be queued on the network and arrive on the plane on the next flight.
+
+A field is added to the MAV_PROTOCOL_CAPABILITY if the COMMAND_STAMPED is supported. Based on the value of this field the ground control station decides if the COMMAND_STAMPED is sent instead of the COMMAND_LONG. Optionally also the user of the Ground Station can decide if he wants to use this new feature.
+
+By separating the COMMAND_LONG and COMMAND_STAMPED and the ability of the user to decline to use the new feature the interference with the existing functionalities is kept to a minimum.
+
+If a command arrives on the vehicle based on the metadata of the COMMAND_STAMPED it is decided if the command is declined or accepted. A command is rejected if:
+ 1. the vehicle_time of a previously received command is larger (this check is only executed if vehicle_time of the received command is larger than 0)
+ 2. the utc_time of a previously received command is larger
```

Comment:
Different systems can have multiple ideas of UTC time, doesn't this end up making the system with a laggy sense of time always rejected?

Reply:
I think that they should be accurate to a second (which is the resolution we can get). As the utc time will be used to distinguish between different flights we could also add there a margin on the plane side. In addition we could also not execute this check if the field is set to 0 which would enable a user to disable the check.

Reply:
I think this is a widely untrue assumption in my experience. I've seen clocks be quite a bit more off, and some systems struggle a lot. As an anecdote a cheap consumer windows 10 machine (that is set to use network time) was off by 17 minutes when I started using it last night and ended up 30 minutes off after 2 hours of usage.

If the time sync was specified as being done over the MAVLink network rather then a system's best idea of UTC (which could be derived from a GPS source (and be wrong until correction data is downloaded), a network source, or internal clocking/user configuration) then this would probably be a lot better. At that point though I don't think there is a large benefit in including this field at all as the `vehicle_time` field is already providing it.

Reply:
Which time did you use? Could I send you a code snippet and you test how well the time is synched to the UTC time?

I see the issue here that is why I will add the feature to disable the check. But by using only the vehicle time there is no way to distinguish the commands from different flights which is also necessary for our use case of satellite communication.

Reply:
This was comparing the system clock to phones which are also doing network time (and github was reporting that I was commenting in the future whenever I viewed a comment I was making).

If you specify that the vehicle's idea of UTC time, as found in `SYSTEM_TIME` is the sync point you now get multiple flights without conflict and only need a single field.